File: src/models/Prescription.js
const mongoose = require('mongoose');
const mongoosePaginate = require('mongoose-paginate-v2');
const baseSchemaFields = require('./BaseSchema');

const prescriptionSchema = new mongoose.Schema({
  ...baseSchemaFields,
  userId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true,
    index: true,
  },
  hospitalName: { type: String, trim: true, maxlength: 255 },
  doctorName: { type: String, trim: true, maxlength: 255 },
  visitDate: { type: Date },
  diagnosis: { type: String, trim: true, maxlength: 1000 },
  gridFsFileId: { type: mongoose.Schema.Types.ObjectId, ref: 'fs.files' },
  originalFileName: { type: String, trim: true, maxlength: 255 },
  mimeType: { type: String, trim: true, maxlength: 255 },
  fileSize: { type: Number, min: 0 },
  sequenceNumber: { type: Number, min: 0 },
  agreement: {
    accepted: { type: Boolean },
    termsAccepted: { type: Boolean },
    termsVersion: { type: String, trim: true, maxlength: 50 },
    acceptedAt: { type: Date },
    appVersion: { type: String, trim: true, maxlength: 50 },
    platform: { type: String, trim: true, maxlength: 50 },
  },
}, { timestamps: true, strict: true });

// Compound index for optimal dashboard counts and pagination
prescriptionSchema.index({ userId: 1, isDeleted: 1, createdAt: -1 });

prescriptionSchema.pre('save', async function(next) {
  if (this.isNew) {
    const Counter = require('./Counter');
    const counter = await Counter.findOneAndUpdate(
      { userId: this.userId, entityType: 'Prescription' },
      { $inc: { seq: 1 } },
      { new: true, upsert: true }
    );
    this.sequenceNumber = counter.seq;
  }
  next();
});

prescriptionSchema.plugin(mongoosePaginate);

module.exports = mongoose.model('Prescription', prescriptionSchema);

File: mobile-app/lib/features/prescriptions/data/models/prescription.dart
class Prescription {
  final String id;
  final String hospitalName;
  final String doctorName;
  final DateTime visitDate;
  final String? diagnosis;
  final String? gridFsFileId;
  final String? originalFileName;
  final String? mimeType;
  final int? fileSize;
  final int? sequenceNumber;
  final DateTime? createdAt;
  final DateTime? updatedAt;

  String get displayId {
    if (sequenceNumber != null) {
      return 'PSCT${sequenceNumber.toString().padLeft(4, '0')}';
    }
    return id.substring(0, 8).toUpperCase();
  }

  Prescription({
    required this.id,
    required this.hospitalName,
    required this.doctorName,
    required this.visitDate,
    this.diagnosis,
    this.gridFsFileId,
    this.originalFileName,
    this.mimeType,
    this.fileSize,
    this.sequenceNumber,
    this.createdAt,
    this.updatedAt,
  });

  factory Prescription.fromJson(Map<String, dynamic> json) {
    return Prescription(
      id: json['_id']?.toString() ?? '',
      hospitalName: json['hospitalName'] ?? '',
      doctorName: json['doctorName'] ?? '',
      visitDate: json['visitDate'] != null ? DateTime.parse(json['visitDate']) : DateTime.now(),
      diagnosis: json['diagnosis'],
      gridFsFileId: json['gridFsFileId'],
      originalFileName: json['originalFileName'],
      mimeType: json['mimeType'],
      fileSize: (json['fileSize'] as num?)?.toInt(),
      sequenceNumber: (json['sequenceNumber'] as num?)?.toInt(),
      createdAt: json['createdAt'] != null ? DateTime.parse(json['createdAt']) : null,
      updatedAt: json['updatedAt'] != null ? DateTime.parse(json['updatedAt']) : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'hospitalName': hospitalName,
      'doctorName': doctorName,
      'visitDate': visitDate.toIso8601String(),
      'diagnosis': diagnosis,
      'gridFsFileId': gridFsFileId,
      'originalFileName': originalFileName,
      'mimeType': mimeType,
      'fileSize': fileSize,
    };
  }
}
